[MRG] Change parse_bids_fname to use full entity names (e.g. 'subject' instead of 'sub') and extract kind and extension

adam2392 opened the pull request:
PR Description
--------------

Closes: #494 
Closes: #497 

See some issues that came up in: #491. 

This will change code that looks like this: 
```Python
        params = parse_bids_filename(bids_basename)
        bids_basename = BIDSPath(subject=params.get('sub'),
                                 session=params.get('ses'),
                                 recording=params.get('rec'),
                                 acquisition=params.get('acq'),
                                 processing=params.get('proc'),
                                 space=params.get('space'),
                                 run=params.get('run'),
                                 task=params.get('task'))
```
to this:

```Python
        params = parse_bids_filename(bids_basename)
        bids_basename = BIDSPath(**params)
```

### Note
@hoechenberger I think you added the `split` kwarg functionality... So the only place that is in currently is `get_bids_entities_from_fname` and having it not supported in `make_bids_basename` and `BIDSPath` makes it hard to work with. Few questions to move forward:

1. Are you okay if I take that out rn and when it is added in later, it's added in everywhere it is needed?
2. If not, any chance you can help me figure out how to add those in, so we can do things like the above code simplifications?

Reference: #426 #418

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

jasmainak commented:
Lovely!! Red lines make my Sunday mornings. 

Can't wait to see the CIs get green :-)

adam2392 commented:
> Lovely!! Red lines make my Sunday mornings.
> 
> Can't wait to see the CIs get green :-)

Me too.. :p 

Looks like in order to do those code simplifications and add `kind` entity I'll need to finish up #492 first.

hoechenberger commented:
> @hoechenberger I think you added the `split` kwarg functionality... So the only place that is in currently is `get_bids_entities_from_fname` and having it not supported in `make_bids_basename` and `BIDSPath` makes it hard to work with. Few questions to move forward:
> 
> 1. Are you okay if I take that out rn and when it is added in later, it's added in everywhere it is needed?

Yeah sure, just go ahead :)

adam2392 commented:
I'm finding that #492 is very nuanced and has **a lot** of edge cases that need to be changed due to the abstract nature of what we previously called `prefix` and `suffix`. Changing `prefix` to `bids_root` is probably the harder part of the two. In addition, it seems like the more I change, the fatter the PR becomes and the slower/harder the reviews. 

I propose adding the small changes to this PR (e.g. suffix -> kind + extension) first and getting it merged in, which will make #492 easier. Changing `suffix` in BIDSPath and make_bids_basename and then the underlying tests is all that is needed to make this PR mergable. I don't intend on changing anything else.

This will then make #492 have less diff.

Is this okay with you @jasmainak @hoechenberger ?

jasmainak commented:
Fine by me if it's okay for @hoechenberger

agramfort commented:
no objection to split the PR and first focus on kind + extension as
replacement to suffix

>

adam2392 commented:
Okay, I think this is good to review once CI is green. I didn't change any of the current underlying functionality w/ that weird function `get_bids_fname()`, but essentially did the following:

1. `suffix` renamed -> `extension` and `kind` in both make_bids_basename and BIDSPath, and altered the underlying tests
2. `parse_bids_fname` renamed -> `get_bids_entities_from_fname` per @hoechenberger suggestion. 
3. some f-string conversions
4. added hardcoded extensions, file kinds to `config.py`.

With 1 & 2, I simplified the codebase a bit with a bunch of red lines.

cc @jasmainak @agramfort

jasmainak commented on mne_bids/path.py:
```diff
@@ -391,33 +437,50 @@ def parse_bids_filename(fname):
+        params[fname_to_entity[key]] = value
+
+    # parse kind last
+    last_entity = fname.split('-')[-1]
+    if '_' in last_entity:
+        kind_and_ext = last_entity.split('_')[-1]
+        if '.' in kind_and_ext:
+            params['kind'] = kind_and_ext.split('.')[0]
+        else:
+            params['kind'] = kind_and_ext
```
why don't you use _get_kind_ext_from_suffix here?

agramfort commented on mne_bids/config.py:
```diff
@@ -61,14 +61,33 @@
-                      'processing', 'run', 'recording',
-                      'space', 'suffix', 'prefix')
+# allowed BIDS path entities
+BIDS_PATH_ENTITIES = ('subject', 'session', 'task', 'run',
```
```suggestion
ALLOWED_PATH_ENTITIES = ('subject', 'session', 'task', 'run',
```

to be consistent

agramfort commented on mne_bids/path.py:
```diff
@@ -391,33 +437,48 @@ def parse_bids_filename(fname):
+                       'acq': 'acquisition', 'run': 'run',
+                       'proc': 'processing', 'space': 'space',
+                       'rec': 'recording', 'split': 'split',
+                       'kind': 'kind'}
```
I would put this next to ALLOWED_PATH_ENTITIES as ALLOWED_PATH_ENTITIES_SHORT

agramfort reviewed (commented):
besides LGTM

hoechenberger commented on doc/api.rst:
```diff
@@ -56,7 +56,7 @@ Path
 
    BIDSPath
-   parse_bids_filename
+   get_bids_entities_from_fname
```
Good idea!

adam2392 commented on mne_bids/config.py:
```diff
@@ -61,14 +61,33 @@
+    'participants', 'scans',
+    'electrodes', 'channels', 'coordsystem', 'events',  # sidecars
+    'headshape', 'digitizer',  # meg-specific sidecars
+    'behav', 'phsyio', 'stim'  # behavioral
```
These are all the "file" kinds in that sense. Idk what to call it still, but they're the characters after the last `_`,  before the extension.

adam2392 replied:
The problem is there isn't a good word for these, so I just used `file_kinds`. Lmk what you think.

agramfort replied:
ok I can live with that

agramfort commented on mne_bids/path.py:
```diff
@@ -540,11 +603,26 @@ def make_bids_basename(subject=None, session=None, task=None,
+    # no matter what the suffix is, kind and extension are last
+    kind = suffix
+    ext = None
+    if suffix is not None:
```
is it actually possible that it is None? 

apparently not. I tried to add:

`assert suffix is not None`

is no test broke. So I think it's safe to avoid this is `if`

agramfort commented on mne_bids/path.py:
```diff
@@ -189,11 +201,14 @@ def get_bids_fname(self, kind=None, bids_root=None, extension=None):
+            kind, extension = _get_kind_ext_from_suffix(new_suffix)
+            bids_fname = self.copy().update(kind=kind, extension=extension)
         else:
             bids_fname = self.copy().update(suffix='{kind}.{extension}')
```
why do I still have this?

agramfort commented on mne_bids/read.py:
```diff
@@ -385,14 +379,11 @@ def read_raw_bids(bids_basename, bids_root, kind=None, extra_params=None,
-    coord_suffix = f'{search_modifier}*_coordsystem.json'
     electrodes_fname = _find_matching_sidecar(bids_fname, bids_root,
-                                              suffix=elec_suffix,
+                                              suffix='electrodes.tsv',
```
can you also change the API of the private functions to avoid dealing with legacy API

adam2392 replied:
Done!

adam2392 commented:
Once the CIs are green, then this feels pretty good to me! Good call on refactoring the `_find_matching_sidecar` too. I feel like the codebase is much more explicit now. 

Later this week... onwards to the challenge of `prefix` -> `root` :(

agramfort reviewed (approved):
@hoechenberger merge if happy

hoechenberger commented:
Great, thanks @adam2392!
